### PR TITLE
node-search refactoring and independent testing

### PR DESCRIPTION
**Pull request by Raulp8**
- refactors node-search, now in separate package
- adds independent testing for node-search

### CODE REVIEW

File: search/node-search.go [hunk 1]
```diff
@@ -26,8 +28,19 @@ type keyValue struct {
...
+	if err != nil {
+		return nil, err
+	}
+	marshErr := yaml.Unmarshal(data, &node)
+	return &node, marshErr
+}
+
 // Given a path in a yaml file return the node at the end of the path
-func findNode(node *yaml.Node, path ...string) (*yaml.Node, error) {
+func FindNode(node *yaml.Node, path ...string) (*yaml.Node, error) {
```
Comment (skeid21) [inline search/node-search.go:43]: There are a few error conditions in this function that we should probably test.

File: search/node-search_test.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	"testing"
+)
+
+func performSearch(t *testing.T, filePath string, yamlPaths ...[]string) {
+	baseNode, parseNode := MakeNode(filePath)
+	if parseNode != nil {
+		t.Fatalf(parseNode.Error())
+	}
+	for _, yamlPath := range yamlPaths {
+		_, searchError := FindNode(baseNode.Content[0], yamlPath...)
```
[the remark this answers is not in the document]
Comment (Raulp8) [reply search/node-search_test.go:27]: see below comment on what we expect of response

File: search/node-search.go (hunk 1 above)
Comment (Raulp8) [reply search/node-search.go:43]: now testing to make sure that findKey( before FindNode) returns node with matching string as last item in path and that if the last item in path is an index, the returned node is a sequence and the index is in the bounds of the sequence

File: search/node-search_test.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		if searchError != nil {
+			t.Errorf(searchError.Error())
```
[the remark this answers is not in the document]
Comment (Raulp8) [reply search/node-search_test.go:34]: important assertations moved to test methods

File: search/node-search.go [hunk 4]
```diff
@@ -26,8 +28,31 @@ type keyValue struct {
 	value *yaml.Node
 }
 
+type Seeking int
+
+// Indicates to search what component of key-value pairing findcomponent should return
```
Comment (AlanPeters) [inline search/node-search.go:33]: match casing for findComponent and is 'search' still accurate in this comment

File: search/node-search.go [hunk 5]
```diff
@@ -58,11 +83,16 @@ func MapKeyValuePairs(content []*yaml.Node) map[string]keyValue {
...
 // Helper in which path aligns with matching key node and needs
 // further to determine further traversal
-func resolveMatchingPath(key *yaml.Node, val *yaml.Node, path ...string) (*yaml.Node, error) {
+func resolveMatchingPath(key *yaml.Node, val *yaml.Node, seeking Seeking, path ...string) (*yaml.Node, error) {
 	if len(path) == 1 {
-		return key, nil
+		switch seeking {
+		case VALUE:
+			return val, nil
+		default:
```
Comment (AlanPeters) [inline search/node-search.go:91]: Not sure what is standard in Go, but should this be the case for KEY? Do we need a default in Go? or should default return an error?

File: search/node-search.go [hunk 6]
```diff
@@ -26,8 +28,31 @@ type keyValue struct {
...
+	var node yaml.Node
+	data, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+	marshErr := yaml.Unmarshal(data, &node)
+	return &node, marshErr
+}
+
+func FindKey(node *yaml.Node, path ...string) (*yaml.Node, error) {
```
Comment (AlanPeters) [inline search/node-search.go:50]: Is the return always either the key, or an error. Is the key always a string, or int for sequence? And would the key always be the last part of the path if it exists?

**Review by AlanPeters — COMMENTED** [review]
Could we generalize the node search to only return the found node? Do we ever use the returned Key, or just check for existence?

File: search/node-search.go (hunk 1 above)
Comment (skeid21) [reply search/node-search.go]: The more detailed success test is awesome.  Verifying all output for a test is a good practice.  But, there are still a number of error states that should likely be tested.

File: search/node-search_test.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+)
+
+func performSearch(t *testing.T, filePath string, seeking Seeking, yamlPaths ...[]string) []*yaml.Node {
+	var foundNodes []*yaml.Node
+	baseNode, parseNode := MakeNode(filePath)
+	if parseNode != nil {
+		t.Fatalf(parseNode.Error())
+	}
+	for _, yamlPath := range yamlPaths {
+		node, searchError := findComponent(baseNode.Content[0], seeking, yamlPath...)
```
Comment (skeid21) [inline search/node-search_test.go:32]: Can we test through the public method FindKey?

File: search/node-search_test.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+func getLastTokensFromPaths(yamlPaths ...[]string) []string {
+	var lastStrings []string
+	for _, token := range yamlPaths {
+		lastStrings = append(lastStrings, token[len(token)-1])
+	}
+	return lastStrings
+}
+
+// Assert that known paths are reported in Search
+func TestHardcodedPaths(t *testing.T) {
```
Comment (skeid21) [inline search/node-search_test.go:63]: Can we think of a better name.  The term Hardcoded implies that we have specific values in business logic our code which are coded.  That doesn't seem to be the what we are testing here.

File: search/node-search_test.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+				tt.Errorf("len(foundNodes) != len(yamlPaths), wanted: %d got: %d", len(trial.yamlPaths), len(foundNodes))
+			}
+			lastTokens := getLastTokensFromPaths(trial.yamlPaths...)
+			for i := 0; i < len(foundNodes); i++ {
+				foundNodeKey := foundNodes[i].Value
+				ExpectedPathToken := lastTokens[i]
+				if ind, numberConversionErr := strconv.Atoi(ExpectedPathToken); numberConversionErr == nil {
+					originalPath := trial.yamlPaths[i]
+					SequenceNode := getSequenceNode(tt, trial.filePath, originalPath)
+					if err := validateIndex(SequenceNode, ind); err != nil {
```
Comment (skeid21) [inline search/node-search_test.go:118]: Should this be an input on the test seeing as we are engineering the test case?

File: search/node-search.go (hunk 4 above)
Comment (Raulp8) [reply search/node-search.go]: got rid of seeking

File: search/node-search.go (hunk 5 above)
Comment (Raulp8) [reply search/node-search.go]: got rid of seeking

File: search/node-search.go [hunk 10]
```diff
@@ -37,17 +62,17 @@ func findNode(node *yaml.Node, path ...string) (*yaml.Node, error) {
...
-		return resolveMatchingPath(foundKeyVal.key, foundKeyVal.value, path...)
+	if foundKeyVal, ok := mapKeyValuePairs(node.Content)[path[0]]; ok {
+		return resolveMatchingPath(foundKeyVal.key, foundKeyVal.value, seeking, path...)
 	}
-	return nil, errors.New("unable to find yaml node")
+	return nil, fmt.Errorf("unable to find yaml node %s in %s", path[0], node.Value)
 }
 
-func MapKeyValuePairs(content []*yaml.Node) map[string]keyValue {
+func mapKeyValuePairs(content []*yaml.Node) map[string]keyValue {
```
[the remark this answers is not in the document]
Comment (Raulp8) [reply search/node-search.go]: Alan and I said this make its more readable than iterating through content

File: search/node-search_test.go (hunk 7 above)
Comment (Raulp8) [reply search/node-search_test.go]: now using findkey

File: search/node-search_test.go (hunk 8 above)
Comment (Raulp8) [reply search/node-search_test.go]: renamed to testkeysearch

File: search/node-search_test.go (hunk 9 above)
Comment (Raulp8) [reply search/node-search_test.go]: refactored

File: search/node-search_test.go [hunk 11]
```diff
@@ -0,0 +1,159 @@
+// Copyright 2021 Google Inc. All Rights Reserved.
```
Comment (skeid21) [inline search/node-search_test.go:1]: Much easier and clearer to read.   Well done.

**Review by skeid21 — APPROVED** [review]
LGTM from me.  Alan might want to review some of the changes he suggested.   

Well done, really ended up with a very refined implementation and test suite.

File: search/node-search_test.go [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+			diff := cmp.Diff(foundFilePos,
+				trial.expectedFilePosition)
+			if diff != "" {
+				tt.Error("UnexpectedFilePosition: diff(-want +got):\n", diff)
+			}
+		})
+	}
+}
+
+func TestErrorConditions(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Raulp8) [reply search/node-search_test.go]: now named TestFindKeyErrors

File: search/node-search_test.go [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+func searchKey(t *testing.T, baseNode *yaml.Node, path ...string) (line, col int) {
+	line, col, searchError := FindKey(baseNode, path...)
+	if searchError != nil {
+		t.Fatalf(searchError.Error())
+	}
+	return line, col
+}
+
+// Assert keysearch results in correct file position
+func TestFilePosition(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Raulp8) [reply search/node-search_test.go]: now named TestFindKey

File: search/node-search.go [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+	return &node, marshErr
+}
+
+// Returns the fileposition of the key at the end of path
+func FindKey(node *yaml.Node, path ...string) (line, col int, err error) {
+	node, searchErr := findComponent(node, path...)
+	if searchErr != nil {
+		return line, col, searchErr
+	}
+	line = node.Line
```
[the remark this answers is not in the document]
Comment (Raulp8) [reply search/node-search.go:55]: now declaring line and col in function body

File: search/node-search_test.go (hunk 11 above)
Comment (AlanPeters) [reply search/node-search_test.go:1]: +1 Nice work here!

**Pull request merged**